stan/math/prim/scal/fun/lower_reg_inc_gamma.hpp

```diff
@@ -0,0 +1,194 @@
+              "did not converge.");
+          k++;
+          lgamma_a_plus_n_plus_1 += log(a_plus_n + 1);
+          a_plus_n++;
+        }
+
+        k = 0;
+        a_plus_n = a;  // n begins at zero.
+        lgamma_a_plus_n_plus_1 = lgamma(a+1);
+        while (true) {
+          term = exp(a_plus_n * log_z - lgamma_a_plus_n_plus_1 +
+            log(digamma(a_plus_n + 1)));
+          sum_b += term;
+          if (term <= precision)
+            return (log_z/exp(z)) * sum_a - inv(exp(z)) * sum_b;
+          if (k >= max_steps)
+            domain_error("d_lower_reg_inc_gamma_da", "k (internal counter)",
+              max_steps, "exceeded ", " iterations, gamma_p(a,z) gradient (a) "
+              "did not converge.");
+          k++;
+          lgamma_a_plus_n_plus_1 += log(a_plus_n + 1);
+          a_plus_n++;
+        }
+      }
+    };
+
+
+    template <>
+    struct lower_reg_inc_gamma<2> {
+      template <int ARG, typename T1, typename T2>
+      static
+      typename boost::math::tools::promote_args<T1, T2>::type
+      grad(T1 a, T2 z, double precision = 1e-10,
+                                 int max_steps = 1e5) {
+        typedef typename boost::math::tools::promote_args<T1, T2>::type TP;
+
+        if (is_nan(a)) return std::numeric_limits<TP>::quiet_NaN();
+        if (is_nan(z)) return std::numeric_limits<TP>::quiet_NaN();
+        double tg = tgamma(a);
+        double dig = digamma(a);
+        return -stan::math::grad_reg_inc_gamma(a, z, tg, dig,
+          max_steps, precision);
+      }
+    };
+
+    template <>
+    struct lower_reg_inc_gamma<3> {
+      template <int ARG, typename T1, typename T2>
+      static
+      typename boost::math::tools::promote_args<T1, T2>::type
+      grad(T1 a, T2 z, double precision = 1e-10,
+                                 int max_steps = 1e5) {
+        typedef typename boost::math::tools::promote_args<T1, T2>::type TP;
+
+        if (is_nan(a)) return std::numeric_limits<TP>::quiet_NaN();
+        if (is_nan(z)) return std::numeric_limits<TP>::quiet_NaN();
+
+        using std::exp;
+        using std::log;
+        using std::pow;
+
+        T2 log_z = log(z);
+
+        int k = 0;
+        TP sum_a = 0.0;
+
+        TP a_plus_n = a;  // n begins at zero.
+        TP lgamma_a_plus_n_plus_1 = lgamma(a+1);
+        TP term;
+        while (true) {
+          term = exp(a_plus_n * log_z - lgamma_a_plus_n_plus_1);
+          sum_a += term;
+          if (term <= precision)
+            break;
+          if (k >= max_steps)
+            domain_error("d_lower_reg_inc_gamma_da", "k (internal counter)",
+              max_steps, "exceeded ", " iterations, gamma_p(a,z) gradient (a) "
+              "did not converge.");
+          k++;
+          lgamma_a_plus_n_plus_1 += log(a_plus_n + 1);
+          a_plus_n++;
+        }
+
+        int n = 1;
+        TP sum_b = pow(z, a) * digamma(a + 1) / tgamma(a + 1);
```

Comment:
This looks dangerous---tgamma is very very prone to overflow, being basically a factorial.  Any way to work this on the log scale so you just the log of `sum_b` instead?  I guess you use it directly in the return, so it doesn't look like it.